### ISSUE TITLE
weighted voronoi diagrams

### ISSUE THREAD

**Issue by Sinasta** [nortikin/sverchok#4405]
I would like to know if there is any way to divide a surface into a weighted voronoi. More precisely i would like to have control over the surface of the different cells which could maybe be done by adding a weight or radius to the site vertex. Would this be possible somehow?

**Comment by summereasy** [comment]
Sorry, I am a bit confused. 
So, you have a dataset for each of every vertex of a Surface, and you want to inject those data into [Vertex Weights](https://docs.blender.org/manual/en/latest/modeling/meshes/properties/vertex_groups/vertex_weights.html) or something else, am I right?

**Comment by Sinasta** [comment]
I am sorry if i did not state my question correctly. As far as i understood the Voronoi lines are drawn be calculating the mean distance between each point. What i would like to be able to is to have an influance over the weight ( or in other word radius) of each of those points. In application this would mean if all the points have the same weight(radius) except one with a bigger weight, the created voronoi cell's surface of this point would be bigger than if it would have the same weight as the others.

![image](https://user-images.githubusercontent.com/64699866/156365764-c31c839b-c627-4df1-8434-dd4783230bcc.png)

**Comment by Sinasta** [comment]
I guess this Wikipedia article describes it better:
https://en.wikipedia.org/wiki/Weighted_Voronoi_diagram

**Comment by Sinasta** [comment]
or this:
https://en.wikipedia.org/wiki/Power_diagram
https://github.com/Kcnarf/d3-weighted-voronoi

**Comment by Sinasta** [comment]
![image](https://user-images.githubusercontent.com/64699866/156366833-76b6863b-a3aa-4b97-a84f-2db6f7788ca3.png)

**Comment by summereasy** [comment]
Oh, I see. Thanks for the information.

So, based on what you've stated above, I think you should take a look at [Voronoi on Mesh](https://github.com/nortikin/sverchok/blob/master/docs/nodes/spatial/voronoi_on_mesh.rst) Node. 

Does that fit your needs?

**Comment by Sinasta** [comment]
Yes i am using Voronoi on mesh (well Voronoi on solid because Voronoi on mesh creates overlapping faces sometimes). How are you able to do weighted Voronoi diagrams with the Voronoi on mesh Node. Maybe I am missing something.

**Comment by Sinasta** [comment]
This here is probably the closest example to what i am looking for:
https://youtu.be/wKfIddiuG58?t=75
https://github.com/DanielAbalde/PowerDiagrams
![image](https://user-images.githubusercontent.com/64699866/156388473-1e561209-e3d2-49e8-bbec-266c9442a256.png)

**Comment by summereasy** [comment]
OK. I see. I didn't look closely. I used to think the "Voronoi on Mesh" node has controls over Weight, But clearly, it doesn't. 
So...  I think the best way to do it may be the "sn-lite" node : )

**Comment by Durman** [comment]
Here is my solution. The idea is to use slice of 3D dimensional Voronoi diagram with Z dimension as weight.

[test.zip](https://github.com/nortikin/sverchok/files/8175108/test.zip)

![image](https://user-images.githubusercontent.com/28003269/156500032-2ad5e3ac-b820-4015-92c8-cdbba10faa9b.png)

**Comment by Sinasta** [comment]
Thank you very much for your answer. Coincidently i found out yesterday that a unweighted Voronoi of N-dimensions is a weighted diagram of N-1 dimension which makes a lot of sens. Somehow it would still be nice if weights could be implemented in the Voronoi node (But this would probably need a switch from Scipy to Cgal). Anyhow the 2d to 3d will do the trick for now. This will only get a problem as soon as somebody needs a weighted 3D Voronoi.

**Comment by portnov** [comment]
Scipy supports n-dimensional voronoi. Just need to try...

**Comment by satabol** [comment]
> Yes i am using Voronoi on mesh (well Voronoi on solid because Voronoi on mesh creates overlapping faces sometimes). 
Hi @Sinasta. I fix nodes "Voronoi on Mesh" and "Voronoi on Solid". Can you try them if you are using sverchok? You can use any configuration of sites: lines, planes without restrictions.

**Comment by satabol** [comment]
Update for the future:
http://www.qhull.org/
![image](https://github.com/nortikin/sverchok/assets/14288520/ad95d78a-7b4d-4588-a1af-74c91e6fa3e2)
So it is not possible realize weighted Voronoi with scipy.spatial.Voronoi. Need something another or some tricks...
